datafusion/core/src/optimizer/projection_push_down.rs

```diff
@@ -172,16 +172,7 @@ fn optimize_plan(
                 _execution_props,
             )?;
 
-            let new_required_columns_optimized = new_input
-                .schema()
-                .fields()
-                .iter()
-                .map(|f| f.qualified_column())
-                .collect::<HashSet<Column>>();
-
-            if new_fields.is_empty()
-                || (has_projection && &new_required_columns_optimized == required_columns)
```

Comment:
I wonder if the issue is that this code is just checking column names to detect a reorder. The real check, as exposed in your reproducer, is that the the column names are the same *AND* that the expressions are only column references or constants... In the case of expressions (like in your reproducer) equal names is not sufficient

Reply:
That makes sense. I'll give it a try tomorrow unless you get to it first!